### PR TITLE
Preparing the foundation for better element IDs

### PR DESCRIPTION
**Pull request by micmarty-deepsense**
Part one of the issue described here: https://github.com/Unstructured-IO/unstructured/issues/2461

It does not change how hashing algorithm works, just reworks how ids are assigned:
> Element ID Design Principles
> 
> 1. A partitioning function can assign only one of two available ID types to a returned element: a hash or UUID.
> 2. All elements that are returned come with an ID, which is never None.
> 3. No matter which type of ID is used, it will always be in string format.
> 4. Partitioning a document returns elements with hashes as their default IDs.

Big thanks to @scanny for explaining the current design and suggesting ways to do it right, especially with chunking.

Here's the next PR in line: https://github.com/Unstructured-IO/unstructured/pull/2673

### CODE REVIEW

File: test_unstructured/documents/test_elements.py
Comment (micmarty-deepsense) [inline test_unstructured/documents/test_elements.py:380]: I think it's better than doing this:
```
elements = assign_hash_ids(elements)
```
which has more logic underneath, and would nee to be imported

File: unstructured/documents/email_elements.py
Comment (scanny) [inline unstructured/documents/email_elements.py:31]: I think this line goes away because `Text` takes care of adding this attribute (self.text) and setting its value when you call `super()` below.

File: test_unstructured/documents/test_elements.py
Comment (scanny) [reply test_unstructured/documents/test_elements.py:389]: Yeah, this one is borderline I think. We should definitely not use `documents.elements.assign_hash_ids()` because this is not its intended use and is not under test here.

But it might not hurt to add a test-utility function to `test_unstructured.unit_utils` to do this. Unfortunately the name that really makes sense to me for that is `assign_hash_ids()`, which could be a little confusing, but still, I think that's probably worthwhile because this same code passage appears a dozen times in this PR and maybe more in future. Then instead of three new lines each time it would just be:
```python
elements = assign_hash_ids([Title("Lorem"), Text("Lorem Ipsum")]
```

I think that's worth doing. We can adjust the name if we really don't like it. `using_hash_ids()` could be a good name, or maybe `with_hash_ids()`.

Or we might prefer to adjust the name of the original to include the fact that it's mapping metadata ids. Maybe something like `assign_and_map_hash_ids()`.

**Review by scanny — APPROVED** [review]
LGTM. A few comments to attend to before merging but approving in advance.

Not sure if Crag wants to take another look before merging, probably worth asking.

File: test_unstructured/documents/test_elements.py
Comment (micmarty-deepsense) [reply test_unstructured/documents/test_elements.py]: done

File: test_unstructured/documents/test_elements.py
Comment (micmarty-deepsense) [reply test_unstructured/documents/test_elements.py]: good point, I'll try to not forget about that in the future. 
test name updated - done

File: unstructured/documents/elements.py
Comment (qued) [inline unstructured/documents/elements.py:694]: If the purpose of the `NoID` class is to be a placeholder indicating we want the `Element` to self-assign an id, do we need the class at all? Could we just use `None` instead (i.e. `element_id` could have type `Optional[str]`)? Or do we need `NoID` for backward compatibility with code out in the wild (in which case maybe we could deprecate)?

**Review by scanny — CHANGES_REQUESTED** [review]
Don't merge yet, we're rethinking our approach on assigning element-ids and in particular the idea of having `element.metadata.parent` as an object reference when the element is in-memory and `element.metadata.parent_id` only appearing in serialized forms (dict, JSON).

File: unstructured/documents/elements.py
Comment (scanny) [reply unstructured/documents/elements.py:694]: I think that's an excellent question @qued. I'm inclined to agree that we should get rid of `NoID` and fix up whatever internal calls we have in a partitioner that might use it.

It doesn't add any new information, at least not anymore. I think at one time you could "pre-select" the "`id_type`" by specifying either `UUID()` (not the stdlib `uuid.UUID`) or `NoID()`, but that's no longer the case and I think this should go away now and just use `None` as the default.

**Review by scanny — CHANGES_REQUESTED** [review]
~Hi Mike, I'm a little confused. I thought this PR was to _prepare_ for better element-ids, but not to actually implement them yet. That the bit that changed the hash algorithm to include page_number and page_sequence_number was going to be separate (because that's what triggers all the ingest-test changes).~

~Did we change our mind on that?~

**Review by scanny — APPROVED** [review]
Hmm, okay, nevermind. I mistook something else for a change to the hash algorithm.

LGTM :)

**Pull request merged**
